conn.go

```diff
@@ -423,7 +423,9 @@ func (c *Conn) loop(ctx context.Context) {
 	for {
 		if err := c.connect(); err != nil {
```

Comment:
connect seems to only return an error when c.shouldQuit is set.  connect already has a reconnect loop as well, so I am not sure this will do what you want it to do.

Reply:
case: zk server startup **AFTER** zk client

Reply:
connect() failed and then *RETURN* loop()

Reply:
Could you help explain this a bit more for me.  I am sure you are right, but I am not picturing the scenario.

Reply:
Also any unittest here would be good.  The logic here is not trivial and a unittest could document the issue as well protect against regressions